src/ComfortPay/Message/ChargeRequest.php

```diff
@@ -18,79 +24,163 @@ public function setTerminalId($value)
             'amount' => $this->getAmount(),
-            'currency' => (new ISOCurrencies())->numericCodeFor(new Currency($this->getCurrency())),
+            'cc' => $this->getCurrency(),
+            'ipspData' => $this->getIpspData(),
             'vs' => $this->getVs(),
             'ss' => $this->getSs(),
+            'e2eReference' => $this->getE2eReference(),
+            'submerchantId' => $this->getSubmerchantId(),
+            'location' => $this->getLocation(),
+            'city' => $this->getCity(),
+            'alpha2CountryCode' => $this->getAlpha2CountryCode(),
         ]);
-        
+
         return $data;
     }
 
     public function sendData($data)
     {
-        if ($this->getTestmode()) {
+        if ($this->getTestmode())
+        {
             if (intval($data['cid']) % 2 == 0) {
-                return $this->response = new ChargeResponse($this, ['transactionStatus' => '02', 'transactionApproval' => '123']);
+                return $this->response = new ChargeResponse($this,
+                    ['transactionId' => $data['transactionId'], 'transactionStatus' => '02', 'transactionApproval' => '123']);
             }
-            return $this->response = new ChargeResponse($this, ['transactionStatus' => '00', 'transactionApproval' => '123']);
+            return $this->response = new ChargeResponse($this,
+                ['transactionId' => $data['transactionId'], 'transactionStatus' => '00', 'transactionApproval' => '123']);
+        }
+
+        $req = new \stdClass();
+        $req->transactionId = $data['transactionId'];
+        $req->referedCardId = $data['referedCardId'];
+        $req->merchantId = $data['merchantId'];
+        $req->terminalId = $data['terminalId'];
+        $req->amount = $data['amount'];
+        $req->parentTransactionId = $data['parentTransactionId'];
+        $req->cc = $data['cc'];
+
+        $symbols = new \stdClass();
+        $symbols->variableSymbol = $data['vs'];
+        $symbols->specificSymbol = $data['ss'];
+
+        $transactionIdentificator = new \stdClass();
+        $transactionIdentificator->symbols = $symbols;
+        $transactionIdentificator->e2eReference = $data['e2eReference'];
+        $req->transactionIdentificator = $transactionIdentificator;
+
+        if (!empty($data['submerchantId']) && !empty($data['location']) && !empty($data['city']) && !empty($data['alpha2CountryCode'])) {
+            $ipspData = new \stdClass();
+            $ipspData->submerchantId = $data['submerchantId'];
+            $ipspData->location = $data['location'];
+            $ipspData->city = $data['city'];
+            $ipspData->alpha2CountryCode = $data['alpha2CountryCode'];
+            $req->ipspData = $ipspData;
         }
 
+        $request = new \stdClass();
+        $request->req = $req;
+        $request->transactionType = $data['transactionType'];
+
         $client = $this->getSoapClient();
+        $response = $client->doCardTransaction($request);
```

Comment:
The original implementation used this line as a parameter for the `$client`.

```php
$param = new SoapParam($data, 'TransactionRequest');	
$response = $client->doCardTransaction($param);
```

Are you sure you don't need to do something similar? (but with the `CardTransactionRequest`?)

Reply:
I've tested it also with SoapParam, but it doesn't work. SoapParam has been removed from all method calls and works perfect for all.